### PR TITLE
Adding ability to get fields via glob patterns for export handler

### PR DESCRIPTION
**Pull request by justinrsweeney**
Currently the Solr stream and export handlers are not able to use glob patterns for fields, similar to how we can for the select handler. For example, passing `fl=*` or `fl=*_bool` to the export handler will not work.

In order to be able to use the export handler with a schema having dynamic fields, we will need the ability to support glob patterns. This implementation finds fields in the schema that match the given pattern and only includes them for export if they are both `docValues=true` and `useDocValuesAsStored=true`, otherwise they are skipped.

### CODE REVIEW

**Comment by magibney** [comment]
looking at this now; for other reviewers' convenience, it's much clearer to view the diff [ignoring whitespace](https://github.com/cowpaths/fullstory-solr/pull/90/files?w=1).

**Comment by magibney** [comment]
Might it be preferable to do this in two passes?:
1. expanding glob fields to a final list of field names
2. use the expanded final list of field names in place of `fields`.

This would allow leaving `writers` as a pre-sized array and would remove the extra indentation (neither of these being terribly important); but more importantly: with the PR in its current state I think that if two glob patterns match the same field that field might get double-added? There's more than one way to address this, but the two-pass approach is the simplest I can think of, and I like that it yields a minimal diff.

**Comment by justinrsweeney** [comment]
> Might it be preferable to do this in two passes?:
> 
> 1. expanding glob fields to a final list of field names
> 2. use the expanded final list of field names in place of `fields`.
> 
> This would allow leaving `writers` as a pre-sized array and would remove the extra indentation (neither of these being terribly important); but more importantly: with the PR in its current state I think that if two glob patterns match the same field that field might get double-added? There's more than one way to address this, but the two-pass approach is the simplest I can think of, and I like that it yields a minimal diff.

Yeah, I thought about doing that but didn't want to introduce another loop, it is probably worth it for clarity of code and your point about avoiding dupes. The minimal diff will be nice too depending on if we push this upstream or not. I'll make that change.

File: solr/core/src/java/org/apache/solr/handler/export/ExportWriter.java [hunk 1]
```diff
@@ -828,4 +829,60 @@ public String getMessage() {
...
+   * Create a list of schema fields that match a given glob pattern
+   *
+   * @param fieldPattern the glob pattern to match
+   * @param searcher an index search to access schema info
+   * @param fieldsProcessed the set of field names already processed to avoid duplicating
+   * @return a list of fields that match a given glob pattern
+   */
+  private List<SchemaField> getGlobFields(
+      String fieldPattern, SolrIndexSearcher searcher, Set<String> fieldsProcessed) {
+    List<SchemaField> schemaFields = new ArrayList<>();
```
Comment (magibney) [inline solr/core/src/java/org/apache/solr/handler/export/ExportWriter.java:874]: nit (please feel free to ignore!):

Why not just make this method void and pass `expandedFields` as an arg to be populated? Avoid creating a new list and iterating it -- and we're already doing similar with `fieldsProcessed` ...

File: solr/core/src/java/org/apache/solr/handler/export/ExportWriter.java [hunk 2]
```diff
@@ -828,4 +829,60 @@ public String getMessage() {
...
+          if (fieldsProcessed.add(field)) {
+            expandedFields.add(searcher.getSchema().getField(field));
+          }
+        }
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    return new ArrayList<>(expandedFields);
```
[the remark this answers is not in the document]
Comment (justinrsweeney) [reply solr/core/src/java/org/apache/solr/handler/export/ExportWriter.java]: Good catch, missed fixing that in some cleanup. Thanks!

**Review by magibney — APPROVED** [review]
LGTM!

**Comment by patsonluk** [comment]
😬 sorry i missed this PR until now, just a really minor thing it appears that `wildcardMatch` can handle both * and ? and glob handling in Solr fl does the same, perhaps we can allow both ? and * in https://github.com/cowpaths/fullstory-solr/pull/90/files#diff-ef9f0e392f110adf48d6ea5bd773b911416bc7937840297937e44e4f97116999R849

Anyway, super minor, I think people usually use `*` anyway 😓

**Comment by justinrsweeney** [comment]
> 😬 sorry i missed this PR until now, just a really minor thing it appears that `wildcardMatch` can handle both * and ? and glob handling in Solr fl does the same, perhaps we can allow both ? and * in https://github.com/cowpaths/fullstory-solr/pull/90/files#diff-ef9f0e392f110adf48d6ea5bd773b911416bc7937840297937e44e4f97116999R849
> 
> Anyway, super minor, I think people usually only use `*` anyway 😓

Np, that's a good thought. I'm going to look to upstream this change anyway so I'll leave as is for now on our fork and work through that as part of upstreaming it.

**Pull request merged**